### DISCUSSION TITLE
/2021/06/26/angular/

### DISCUSSION THREAD

**Discussion by BookOfCooks** [WorldMaker/blog.worldmaker.net#33] [Posts]
I read your [blog post](https://blog.worldmaker.net/2021/06/26/angular/#suggestions-for-a-better-angular-project-gawky) and I think it's amazing and encapsulates everything I currently hate about Angular. Especially the manual .subscribe() and using Subject/ReplaySubject/BehaviorSubject escape hatches used.

P.S: Angular signals may bring us to a Zoneless angular, but that'll be in a long time.

All that out of the way, I checked out angular-pharkas and I have to say that's very complicated and verbose for its purpose. I imagine something else could be done that's similar. Here's how I approach it.

```
@Component({
    selector: 'app-component',
    standalone: true,
    imports: [
      CommonModule,
      FormsModule,
      ButtonElementDirective // this is my own directive, along with ButtonElement class
    ],
    template: `
      <button [buttonElement]="purchaseBtn">Purchase</button>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
  })
  export class SpecialComponent {
    purchaseBtn = new ButtonElement();
    $purchaseRequest = this.purchaseBtn.click$.pipe(
      // throttle, switchMap(), shareReplay(), etc
    )
  }
```

Now here's the implementation of the directive:
```
import { Directive, HostListener, Input } from '@angular/core';
import { Subject } from 'rxjs';

@Directive({
  selector: '[buttonElement]',
  standalone: true,
})
export class ButtonElementDirective {
  @Input()
  buttonElement: ButtonElement = new ButtonElement();

  constructor() {}

  @HostListener('click', ['$event'])
  onSubmit(event: PointerEvent) {
    this.buttonElement.click$.next(event);
  }
}

export class ButtonElement {
  click$ = new Subject<PointerEvent>();
}
```
I know, I probably should find a way to use **fromEvent()** instead of **click$.next()**, but it's a lazy implementation and angular complicates things.

**Comment by WorldMaker** [comment]
Hi, thanks for the feedback. Looks like I missed a bunch of GitHub emails at some point, so I just now saw this.

That's certainly an interesting approach. I think there are some things that stand out to me that were important to me when I was writing Pharkas:

1. **Full** Lifecycle management: your Component and Directive are missing `ngOnInit` and `ngOnDestroy`. At least one, if not both should implement that (to manage Subscriptions/Subjects). On top of that, the `@Input buttonElement` could be rebound at any time, it might be late bound just before `ngOnInit`, it might change at the parent Component's whim, it could be unbound late (just after `ngOnDestroy`). It was important to me with Pharkas to handle *all* changes to any Input, including teardown at `ngOnDestroy`, for the full lifecycle of the Component. What does it mean if you change the `buttonElement` input? (In Pharkas Inputs that might even be a console warning in development, it might indicate a garbage leak.)
2. Avoiding `Subject` leakage everywhere possible. `click$ = new Subject<PointerEvent>().asObservable()` in `ButtonElement` at least would avoid a possible leak, but then you can't use that class as an input into your Directive like that because the Directive wouldn't have access to `next()` (which is one reason I try to avoid Subjects as APIs in the first place and try to entirely encapsulate them).
3. I don't like Directives and generally try to avoid them. I also don't like Angular's use of Experimental Decorators and try to avoid most of those, too. I can't imagine how many Directives you would need in a library to support even basic HTML elements if you were trying to create wrappers for everything that you might want to be Observable. (You are right that I would reach for `fromEvent()` long before I tried to use `@HostListener`.)

You are right that Pharkas turned out somewhat verbose, certainly more verbose than I had intended/hoped for. After using it a while, I feel like the verbosity is actually less than most Angular scaffolders produce or most Angular components use in general (and most people copy and paste), especially once you consider that with Pharkas you never write another `ngOnInit` or `ngOnDestroy` ever again. Certainly, a problem to some sorts of Angular users is that Pharkas just isn't very amenable to a scaffolding-based approach. (Maybe there's an opening for a snippet library of some sort?) I certainly *tried* to make it less verbose. I had been greatly hoping to meta-program things a little further than I was able to succeed with. I hit the wall at trying to metaprogram Angular's hideous decorators and Typescript's "please don't use this in Production" experimental decorator output. I couldn't find an easy way to attach decorators without reverse engineering that mess, and I didn't want to reverse engineer that mess because Angular shouldn't have been using it in the first place. At one point I was trying for something that looked even more like React Hooks where the various `use*` functions would auto-create things like `@Input` properties but setting a property to be `@Input` in current Angular is a ton more work than just calling something like `MarkInput(this.prop)`. Angular doesn't offer at all a raw, pure JS API for most of what it uses Decorators for today (which I think is kind of ridiculous).

**Comment by WorldMaker** [comment]
Probably also fun to note, with reference to full lifecycle management, that your example shows `ChangeDetectionStrategy.OnPush` but doesn’t implement it. To do that you need a `ref : ChangeDetectorRef` and to call `ref.detectChanges()`. That’s also something I’m proud that the “verbosity” of Pharkas buys as a “free” win: you have to include some boilerplate to get Angular to DI the `ref`, but all the `ref.detectChanges()` calls are handled automatically by Pharkas including by default throttling them to “Request Animation Frame time”, making for IMNSHO the easiest way to get some crazy smooth template updates. (I’m quite proud of the “real-time noisy data updates” stuff in the demo site for Pharkas. Especially because it is simple fake data I don’t think many realize how hard it is to get that smooth in baseline Angular for how much the code looks simple and unaware of the noisiness. Watching the RxJS-Spy logs for it is a treat for this salty RxJS performance engineer.)

**Comment by BookOfCooks** [comment]
Several months now from writing this post, I've come to all the same conclusions (I care less about the decorators being experimental, because I've come to dislike decorators in general.)

Signals (I'm sure you've heard of them) are a comfortable compromise for me (losing throttle, switchmap, etc).

My original idea when writing this post few months ago would be to have RxJS integrated into a language as syntax ([reddit post here](https://www.reddit.com/r/webdev/comments/12c2u73/building_a_language_around_rxjs/)).

It's still an idea I'd like to throw my hand at one day, when I got time.

**Comment by WorldMaker** [comment]
Yeah, I mostly complain about decorators being experimental because I dislike them and the experimental flag should have been enough to discourage more devs from using them in production.

Some of your syntax ideas would get solved if the [Pipeline Operator proposal](https://github.com/tc39/proposal-pipeline-operator) ever makes it past Stage 2.

**Comment by BookOfCooks** [comment]
Yes, the pipe operator would solve a myriad of syntax issues, but I wanted to try my hand at something I remember reading from you on ycombinator. I apologize in advance for any imposed nightmares.

> I really can't stop thinking about an Observable-only AOT compiled template language-based framework. You could achieve a compile-time version of React's Suspense framework incredibly cheaply (from a user developer standpoint) entirely in the way you compile-time compose, schedule, and subscribe Observables in a template and its child templates. So much of the work that React's Suspense has to do at runtime, such as scheduling things to arrive at requestAnimationFrame() time, debouncing/throttling updates, coordinating (composing) related updates in child components, all have simple RxJS operator equivalents, and you could declare all of that simply in templates leaving the compiler to do the final RxJS operator compositions and most users of the templates to not have any idea how complicated some of those final compiled compositions may be, but directly gaining the benefits of them (such as cleanly debounced AOT-compiled DOM updates at requestAnimationFrame time; complexly coordinated multi-component update patterns). It's an amazing sounding framework in my head. Very worried I at least need to build a prototype of it now or it will continue to haunt my dreams.

**Comment by WorldMaker** [comment]
Oh yeah, that was definitely from the thought-process that directly led to Pharkas. In Pharkas' case it mostly just moved into the "bind" language (`bind` versus `bindImmediate`). I haven't had reason to write a `bindSuspense` or anything like that, but that would be the next most obvious thing missing from my own comment you quoted and you can build it yourself out of an ordinary `combineLatest` and a bunch of `skipUntil` operators. (It's an interesting idea. I decided to track it in WorldMaker/angular-pharkas#52)

If I had better access to overriding Angular's template engine, I would have moved more of the boilerplate more directly into the HTML templates, yes, but I think the end result would have looked very similar where unadorned template bindings (in Angular, things like `[something]="anObservable"` or `{{ otherObservable }}`) would default to nice schedules (Pharkas `bind` schedules to `requestAnimationFrame` time, as it is as nice as I thought it would be in the comment you quoted). You could in theory take something like the "piped" approach for Immediate bindings (`[something]="anObservable | immediate"` or `{{ otherObservable | immediate }}`). It's mostly just Suspense that I think would need a bit more Syntax to be nice to work with, and mostly just because you'd probably want a full tag for suspense and you also might want that tag to auto "slurp" all the Observables bound in templates contained within it.

If I were to start an HTML template engine from scratch (and not use JSX, though I'd probably prefer JSX) I'd probably use the `|>` for a pipeline operator (unlike Angular just using `|`), and maybe support more RxJS operators as "out of the box" pipelining even in templates (`[something]="anObservable |> throttleTime(3000)"`). You wouldn't want to put all of RxJS in templates, and that may end up being too tempting, but some things like throttles/debounces would likely be nice.

**Comment by WorldMaker** [comment]
@BookOfCooks So you might appreciate this. I've had a weird few months and was revisiting some old code that was written in Knockout and thinking I should modernize it, but then realizing that it was sort of elegant as-is and I didn't particular want to modernize it to React or Vue or Svelte or Qwik. I kind of got that urge to "build a template engine from scratch", using some of the lessons from Pharkas and Real/Pure Observables (not compromises like signals or whatever you want to call them).

I'm calling it "Butterfloat". I *just* hit my "MVP" milestones, which was "enough to convert one of my simpler old demo apps written in Knockout plus mostly feature parity with Pharkas". That's currently version 0.3.0. Once I write way more documentation, and maybe decide on a few more of the issues in the issue tracker, I'll may bump it to 1.0 and then figure out if there's a "press release" blog post in me to maybe post somewhere like HN or something.

Here's the most documentation I have written so far: 

https://worldmaker.net/butterfloat/

Unfortunately the diff is not clean at all, but you can sort of see how the Knockout and Butterfloat versions wound up looking somewhat similar in the actual demo app:

https://github.com/WorldMaker/compradprog/pull/6/files

Anecdotally, in this case, with the move to TSX and switching Knockout for Butterfloat (with RxJS dependency) and including rxjs-spy in the current bundle, the (unminified, unzipped) esbuild output only went up 130 kb without yet checking it for missed tree-trimming opportunities (and obviously rxjs-spy is probably not something that should be in a "production" bundle, but is fun to keep around in a demo app like this). That seems reasonable so far.

Butterfloat itself is tiny, and at various points in its history was way more Typescript types and comments than code.

I'd love feedback on it at this early stage. I think I've got something interesting here so far.

**Reply by BookOfCooks** [reply]
I've been away from the JavaScript Landscape for awhile, using Golang and HTMX. I've been in this phase of simplicity with little to no 3rd party dependencies. I don't have much input now that I've been on this break, but I think you're going on the right path by binding events to observables directly.

Other than that, I really don't have much input. Maybe once I'm off the htmx train, I might try butterfly out.

**Reply by WorldMaker** [reply]
Thanks. Yeah, I'm rather proud of how events turned out. That was one of my focuses from the first few sketches in my notes, especially the "circular dependency" problem of wanting to pass them into the component's function but also bind them in the JSX before the function returns. Using a simple 'expando' proxy for that feels like an elegant JS solution and then adding some smart generic types to it makes it almost feel like a smarter dependency injection system than it actually is.

One thing it's also fun to point out that you don't get an easy feel for from the documentation (I don't know how to do that other than maybe add some sort of Playground) is just how *good* the JSX types feel. That was one of my biggest worries going in to the project. React uses a massive multi-thousand line file of apparently hand-maintained types (in `@types/react` on DefinitelyTyped) for JSX. I did not want to copypasta that (like many others do), especially because it could have exponentially blown up into copypasta of copypasta to define the "bind" types. Instead I did a ton of research into Typescript's own (auto-generated from things like MDN metadata) `lib.dom` types. In a couple dozen lines of (meta-)types I managed to get the `lib.dom` types themselves to do the heavy lifting. I think it is a better developer experience than `@types/react` because those types include a lot more documentation comments, including lots of direct MDN links. I probably should PR a version of it to `@types/react` itself, but for now I'm still doing a victory lap at a clever win.
